apps/examples/reboot_reason_test/reboot_reason_main.c

```diff
@@ -0,0 +1,225 @@
+#endif
+#ifdef CONFIG_WATCHDOG
+	printf("\t-Press W or w : Watchdog Reset Test\n");
+#endif
+	printf("\t-Press V or v : Random Value Write-Read Test\n");
+	printf("\t-Press C or c : Clear reboot reason\n");
+	printf("\t-Press R or r : Check previous reboot reason\n");
+	printf("\t-Press X or x : Terminate Tests.\n");
+}
+#ifdef CONFIG_WATCHDOG
+static void watchdog_test(void)
+{
+	int fd;
+	int ret;
+	/* Test for Watchdog Reset */
+	printf("After watchdog reset, Expected Reboot reason is %d\n", REBOOT_SYSTEM_WATCHDOG);
+
+	fd = open(CONFIG_WATCHDOG_DEVPATH, O_RDWR);
+	if (fd < 0) {
+		printf("ERROR: Fail to open %s\n", CONFIG_WATCHDOG_DEVPATH);
+		return;
+	}
+
+	ret = ioctl(fd, WDIOC_SETTIMEOUT, 1000);
+	if (ret != OK) {
+		printf("ERROR: Fail to set watchdog timeout.\n");
+		close(fd);
+		return;
+	}
+
+	ret = ioctl(fd, WDIOC_START, 0);
+	if (ret != OK) {
+		printf("ERROR: Fail to start watchdog.\n");
+		close(fd);
+		return;
+	}
+}
+#endif
+
+#ifdef CONFIG_MM_ASSERT_ON_FAIL
+static void memory_alloc_fail_test(void)
+{
+	struct mallinfo data;
+	int largest_size;
+	int *alloc_ptr;
+	/* Test for Memory Allocation Fail */
+	printf("After Memory Allocation Fail, Expected Reboot reason is %d\n", REBOOT_SYSTEM_MEMORYALLOCFAIL);
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+	data = mallinfo();
+#else
+	mallinfo(&data);
+#endif
+	largest_size = data.mxordblk;
+	/* Try to allocate not available size. */
+	alloc_ptr = (int *)malloc(largest_size + 1);
+
+	/* This line cannot be reached,
+	 * because there will be assert on memory allocation failure from malloc.
+	 */
+	free(alloc_ptr);
+}
+#endif
+
+static void reboot_board(void)
+{
+#ifdef CONFIG_LIB_BOARDCTL
+	printf("Board will be reset automatically.\n");
+	boardctl(BOARDIOC_RESET, 0);
+#else
+	printf("Please reboot the board manually.\n");
+#endif
+}
+
+/****************************************************************************
+ * reboot_reason_main
+ ****************************************************************************/
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int reboot_reason_main(int argc, char *argv[])
+#endif
+{
+	int ch;
+	bool is_testing = true;
+
+	printf("Reboot Reason Test Start!\n");
+
+	while (is_testing) {
+		display_reboot_reason_option();
+		ch = getchar();
+		switch (ch) {
+		case 'H':
+		case 'h':
+			printf("Press HW Reset Button.\n");
+			printf("Expected reboot reason after hw reset is : %d\n", REBOOT_SYSTEM_HW_RESET);
+			/* Wait for pressing the hw reset button. */
+			sleep(5);
+			break;
+#ifdef CONFIG_MM_ASSERT_ON_FAIL
+		case 'M':
+		case 'm':
+			memory_alloc_fail_test();
+			break;
+#endif
+#ifdef CONFIG_WATCHDOG
+		case 'W':
+		case 'w':
+			watchdog_test();
+			break;
+#endif
+		case 'V':
+		case 'v':
+			/* Test for Temp value write-read */
+			printf("Write Test Reboot Reason : %d\n", REBOOT_REASON_TEST);
+			prctl(PR_REBOOT_REASON_WRITE, REBOOT_REASON_TEST);
+			reboot_board();
+			break;
+		case 'C':
+		case 'c':
```

Comment:
When we write random value, it does to board reset directly. Then, when can we clear?

Reply:
Modified the scenario. Thanks.